File: day6/day6_test.py
import dis

def foo(a, len):
    for i in range(len):
        a[i] = i

def bar(a, len):
    result = 0
    if a[len - 1] < 5:
        result = -1
    elif a[len - 1] > 7:
        result = 0
    else:
        result = 1
    return result

N = 8
a = [0] * N
foo(a, N)
print(bar(a, N))

print("\n===== foo function =====")

dis.dis(foo)

print("\n===== bar function =====")

dis.dis(bar)

File: day6/day6_test.cpp
// This example is used to show how the compiler compiles C++ code
// and how the debugger work
//
// Use "g++ -std=c++11 -E day6_test.cpp > day6_test_new.cpp" to see how the preprocessor work
// Use "clang++ -std=c++11 -S -emit-llvm day6_test.cpp" to see the generated LLVM IR (you need clang++)
// Use "g++ -std=c++11 -S -masm=intel -fverbose-asm day6_test.cpp" to see what the generated x86 assembly

#include <iostream>

void foo(int a[], int len) {
    for (int i = 0; i < len; i++)
        a[i] = i;
}

int bar(int a[], int len) {
    int result = 0;

    if (a[len - 1] < 5) {
        result = -1;
    } else if (a[len - 1] > 7) {
        result = 0;
    } else {
        result = 1;
    }
    return result;
}

int main() {
    int N = 8;
    int a[N];
    foo(a, N);

    std::cout << bar(a, N) << '\n';
}
